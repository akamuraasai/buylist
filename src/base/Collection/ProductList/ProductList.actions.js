export const findById = id => ({ type: 'FIND_BY_ID', id });

export const changeField = (field, value) => ({ type: 'CHANGE_FIELD', field, value });

export const apply = item => ({ type: 'APPLY', item });

export const addItem = () => ({ type: 'ADD' });

export const remove = id => ({ type: 'REMOVE', id });

export const toggleFilter = () => ({ type: 'TOGGLE_FILTER' });
