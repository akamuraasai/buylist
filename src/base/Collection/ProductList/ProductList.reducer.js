import { T, cond, propEq } from 'ramda';

const lastId = items => Math.max(...items.map(({ id }) => id))

const applyChanges = state => ({ item }) =>
 (item.id === undefined)
   ? { ...state, list: [...state.list, { ...item, id: lastId(state.list) + 1 }] }
   : { ...state, list: state.list.map(i => i.id === item.id ? item : i) };

const reducer = state => cond([
  [propEq('type', 'APPLY'), applyChanges(state)],
  [propEq('type', 'FIND_BY_ID'), ({ id }) => ({ ...state, item: state.list.find(i => i.id === id) })],
  [propEq('type', 'CHANGE_FIELD'), ({ field, value }) => ({ ...state, item: { ...state.item, [field]: value } })],
  [propEq('type', 'ADD'), () => ({ ...state, item: {} })],
  [propEq('type', 'REMOVE'), ({ id }) => ({ ...state, list: state.list.filter(i => i.id !== id) })],
  [propEq('type', 'TOGGLE_FILTER'), () => ({ ...state, filter: !state.filter })],
  [T, () => state],
]);

const list = [
  { id: 1, unitType: 'kg', name: 'Tomate', price: 3.99, units: 1, completed: false },
  { id: 2, unitType: 'kg', name: 'Cebola', price: 1.99, units: 2.3, completed: false },
  { id: 3, unitType: 'kg', name: 'Alho', price: 4.99, units: 0.4, completed: true },
  { id: 4, unitType: 'un', name: 'Leite', price: 2.49, units: 12, completed: false },
  { id: 5, unitType: 'un', name: 'Bolacha', price: 1.29, units: 5, completed: false },
  { id: 6, unitType: 'kg', name: 'Carne', price: 23.99, units: 0.9, completed: false },
  { id: 7, unitType: 'kg', name: 'Frango', price: 7.99, units: 2.3, completed: true },
  { id: 8, unitType: 'un', name: 'Margarina', price: 4.49, units: 1, completed: false },
];

export default (state = { list, item: {}, filter: false }, action) => reducer(state)(action);
