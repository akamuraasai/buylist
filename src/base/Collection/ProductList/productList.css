.ProductList {
    margin-top: 10px;
}

.ProductList__list {
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
}

.ProductList__item {
    width: 95%;
    height: 74px;
    background-color: #f7faf5;
    margin-bottom: 10px;
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    border-radius: 4px;
    box-shadow: 0px 4px 4px 0px #cbcbcb;
}

.ProductList__item img {
    width: 60px;
    height: 60px;
}

.ProductList__item__image {
    width: 18%;
    margin-left: 2%;
    margin-top: 4px;
}

.ProductList__item__text {
    width: 58%;
    margin-left: 4px;
}

.ProductList__item__text p {
    margin: 0;
}

.ProductList__item__text__title {
    font-size: 16px;
    font-weight: bold;
}

.ProductList__item__price {
    width: 20%;
    margin-right: 2%;
    text-align: center;
    font-size: 15px;
    font-weight: bold;
}

.is-complete {
    background-color: rgba(158, 239, 77, 0.36);
}