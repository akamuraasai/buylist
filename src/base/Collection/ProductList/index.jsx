import React from 'react';
import PropTypes from 'prop-types';

import List from './components/List';
import Modal from '../../Components/Modal';
import './productList.css';

const ProductList = ({ products, filter }) => (
  <div className="ProductList">
    <Modal />
    <List items={products} filter={filter} />
  </div>
);

ProductList.defaultProps = {
  products: [],
};

ProductList.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape()),
};

export default ProductList;
