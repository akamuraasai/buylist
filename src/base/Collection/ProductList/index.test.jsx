import React from 'react';
import { shallow } from 'enzyme';
import ProductList from './index';

describe('ProductList', () => {
  const component = shallow(<ProductList />);

  it('should render without crashing', () => {
    expect(component).toMatchSnapshot();
  });
});
