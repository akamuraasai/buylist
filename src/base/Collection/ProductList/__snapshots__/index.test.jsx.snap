// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProductList
    products={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Modal
          completed={true}
          name="Tomate"
          price={3.59}
          unitType="kg"
          units={3}
/>,
        <List
          items={Array []}
/>,
      ],
      "className": "ProductList",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "completed": true,
          "name": "Tomate",
          "price": 3.59,
          "unitType": "kg",
          "units": 3,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "items": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Modal
            completed={true}
            name="Tomate"
            price={3.59}
            unitType="kg"
            units={3}
/>,
          <List
            items={Array []}
/>,
        ],
        "className": "ProductList",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "completed": true,
            "name": "Tomate",
            "price": 3.59,
            "unitType": "kg",
            "units": 3,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "items": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
